prov/rxm/src/rxm_ep.c

```diff
@@ -781,24 +799,202 @@ rxm_ep_format_tx_res(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
 	if (OFI_UNLIKELY(ret))
 		return ret;
 
-	*tx_entry = rxm_tx_entry_get(&rxm_ep->send_queue);
-	if (OFI_UNLIKELY(!*tx_entry)) {
-		ret = -FI_EAGAIN;
+	ret = rxm_ep_format_tx_entry(rxm_ep, context, count, flags, comp_flags,
+				     *tx_buf, tx_entry);
+	if (OFI_UNLIKELY(ret))
 		goto err;
-	}
-
-	(*tx_entry)->count = count;
-	(*tx_entry)->context = context;
-	(*tx_entry)->flags = flags;
-	(*tx_entry)->tx_buf = *tx_buf;
-	(*tx_entry)->comp_flags = comp_flags | FI_SEND;
 
 	return FI_SUCCESS;
 err:
 	rxm_tx_buf_release(rxm_ep, *tx_buf);
 	return ret;
 }
 
+static inline ssize_t
+rxm_ep_normal_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+		   struct rxm_tx_entry *tx_entry, size_t pkt_size)
+{
+	ssize_t ret = fi_send(rxm_conn->msg_ep, &tx_entry->tx_buf->pkt, pkt_size,
+			      tx_entry->tx_buf->hdr.desc, 0, tx_entry);
+	if (OFI_UNLIKELY(ret)) {
+		if (ret == -FI_EAGAIN)
+			rxm_ep_progress_multi(&rxm_ep->util_ep);
+		else
+			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+				"fi_send for MSG provider failed\n");
+		rxm_tx_buf_release(rxm_ep, tx_entry->tx_buf);
+		rxm_tx_entry_release(&rxm_ep->send_queue, tx_entry);
+	}
+	return ret;
+}
+
+static inline ssize_t
+rxm_ep_alloc_lmt_tx_res(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+			void *context, uint8_t count, const struct iovec *iov,
+			void **desc, size_t data_len, uint64_t data, uint64_t flags,
+			uint64_t tag, uint64_t comp_flags, uint8_t op,
+			struct rxm_tx_entry **tx_entry)
+{
+	struct rxm_tx_buf *tx_buf;
+	struct fid_mr **mr_iov;
+	ssize_t ret;
+
+	/* Use LMT buf pool instead of buf pool provided to the function */
+	ret = rxm_ep_format_tx_res(rxm_ep, rxm_conn, context, (uint8_t)count,
+				   data_len, data, flags, tag, comp_flags,
+				   &tx_buf, tx_entry,
+				   &rxm_ep->buf_pools[RXM_BUF_POOL_TX_LMT]);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+	tx_buf->pkt.hdr.op = op;
+	tx_buf->pkt.ctrl_hdr.msg_id = rxm_txe_fs_index(rxm_ep->send_queue.fs,
+						       (*tx_entry));
+	if (!rxm_ep->rxm_mr_local) {
+		ret = rxm_ep_msg_mr_regv(rxm_ep, iov, (*tx_entry)->count,
+					 FI_REMOTE_READ, (*tx_entry)->mr);
+		if (ret)
+			goto err;
+		mr_iov = (*tx_entry)->mr;
+	} else {
+		/* desc is msg fid_mr * array */
+		mr_iov = (struct fid_mr **)desc;
+	}
+
+	return rxm_rma_iov_init(rxm_ep, &(*tx_entry)->tx_buf->pkt.data, iov,
+				count, mr_iov);
+err:
+	rxm_tx_entry_release(&rxm_ep->send_queue, (*tx_entry));
+	rxm_tx_buf_release(rxm_ep, tx_buf);
+	return ret;
+}
+
+static inline ssize_t
+rxm_ep_lmt_tx_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
+		   struct rxm_tx_entry *tx_entry, size_t pkt_size)
+{
+	ssize_t ret;
+
+	if (pkt_size <= rxm_ep->msg_info->tx_attr->inject_size) {
+		RXM_LOG_STATE(FI_LOG_EP_DATA, tx_entry->tx_buf->pkt, RXM_TX, RXM_LMT_TX);
+		RXM_LOG_STATE(FI_LOG_CQ, tx_entry->tx_buf->pkt, RXM_LMT_TX, RXM_LMT_ACK_WAIT);
+		tx_entry->state = RXM_LMT_ACK_WAIT;
+
+		ret = fi_inject(rxm_conn->msg_ep, &tx_entry->tx_buf->pkt, pkt_size, 0);
```

Comment:
Should we use rxm_ep_inject_send here? It also updates the counters which is missing here.

Reply:
Based on the function name, this is part of a large message transfer.  So this is just using inject to send the packet rather than send.  The counters shouldn't be updated until the entire transfer completes.

Reply:
Yep. Just noticed it. :)